libpromises/mod_custom.c

```diff
@@ -0,0 +1,651 @@
+    PromiseModule_AppendAllAttributes(module, pp);
+    PromiseModule_Send(module);
+
+    // Prints errors / log messages from module:
+    JsonElement *response = PromiseModule_Receive(module);
+
+    const bool valid = HasResultAndResultIsValid(response);
+
+    JsonDestroy(response);
+
+    if (!valid)
+    {
+        // Detailed error messages from module should already have been printed
+        const char *const promise_type = pp->parent_section->promise_type;
+        const char *const promiser = pp->promiser;
+        const char *const filename =
+            pp->parent_section->parent_bundle->source_path;
+        const size_t line = pp->offset.line;
+        Log(LOG_LEVEL_ERR,
+            "%s:%zu: %s promise with promiser '%s' failed validation",
+            filename,
+            line,
+            promise_type,
+            promiser);
+    }
+
+    return valid;
+}
+
+static inline const char *LogLevelToRequestFromModule()
+{
+    // We will never request LOG_LEVEL_NOTHING or LOG_LEVEL_CRIT from the
+    // module:
+    const LogLevel global = LogGetGlobalLevel();
+    if (global < LOG_LEVEL_ERR)
+    {
+        assert((global == LOG_LEVEL_NOTHING) || (global == LOG_LEVEL_CRIT));
+        return LogLevelToString(LOG_LEVEL_ERR);
+    }
+    return LogLevelToString(global);
+}
+
+static PromiseResult PromiseModule_Evaluate(PromiseModule *module, const Promise *pp)
+{
+    assert(module != NULL);
+    assert(pp != NULL);
+
+    PromiseModule_AppendString(module, "operation", "evaluate_promise");
+    PromiseModule_AppendString(module, "log_level", LogLevelToRequestFromModule());
+    PromiseModule_AppendString(module, "promiser", pp->promiser);
+
+    PromiseModule_AppendAllAttributes(module, pp);
+    PromiseModule_Send(module);
+
+    JsonElement *response = PromiseModule_Receive(module);
+    PromiseResult result;
+    const char *const result_str = JsonObjectGetAsString(response, "result");
+
+    if (result_str == NULL)
+    {
+        result = PROMISE_RESULT_FAIL;
+        Log(LOG_LEVEL_ERR,
+            "Promise module did not return a result for promise evaluation (bug in promise module: %s)",
+            module->path);
+    }
+    else if (StringEqual(result_str, "kept"))
+    {
+        result = PROMISE_RESULT_NOOP;
+    }
+    else if (StringEqual(result_str, "not_kept"))
+    {
+        result = PROMISE_RESULT_FAIL;
+    }
+    else if (StringEqual(result_str, "repaired"))
+    {
+        result = PROMISE_RESULT_CHANGE;
+    }
+    else if (StringEqual(result_str, "error"))
+    {
+        result = PROMISE_RESULT_FAIL;
+        Log(LOG_LEVEL_ERR,
+            "An unexpected error occured in promise module (bug in promise module: %s)",
+            module->path);
+    }
+    else
+    {
+        result = PROMISE_RESULT_FAIL;
+        Log(LOG_LEVEL_ERR,
+            "Promise module returned unacceptable result: '%s' (bug in promise module: %s)",
+            result_str,
+            module->path);
+    }
+
```

Comment:
Should do `cfPS()` somewhere around here.

Reply:
@vpodzime Could you add a comment to `eval_context.h` explaining how to use `cfPS()`? It is quite cryptic to me.

Reply:
https://tracker.mender.io/browse/CFE-3442

Reply:
I did my best to add cfPS's

Reply:
> @vpodzime Could you add a comment to `eval_context.h` explaining how to use `cfPS()`? It is quite cryptic to me.

I will do my best!

Reply:
https://tracker.mender.io/browse/CFE-3442

Reply:
@vpodzime https://tracker.mender.io/browse/CFE-3442